design.md

```diff
@@ -0,0 +1,75 @@
+* `Newsupervisor()` - Returns a new instance of a supervisor
+* `Add(cmd string, args []string, opts...Options)` - Adds a new job without running it.
+* `Start(id string)` - Starts a previously added job.
+* `Stop(id string)` - Stops a running job by sending the job's process group a `SIGTERM` signal.
+* `Signal(id string, sig syscall.Signal)` - Sends an os signal to a job's process group.
+* `Run()` - Run the supervisor and listen for incoming job requests.
+* `WatchOutput(id string, c chan *pb.JobOutput)` - Add an output to channel to subscribe to process output. All previous output, if any, from the process is replayed before sending new messages to the channel.
+* `UnWatchOutput(id string, c chan *pb.JobOutput)` - Unsubscribe to the process output.
+* `WatchStatus(id string, c chan *pb.JobStatus)` - Subscribe to state changes for the job's process on the given channel.
+* `WatchStatus(id string, c chan *pb.JobStatus)` - Unsubscribe to state changes for job's process.
+
+### Job API
+`Job` manages a single process and it's output and status updates.
+* `NewJob(p *Process)` - Returns a new instance of a job.
+* `Start()` - Start the process assigned the job.
+* `Stop()` - Stop the process assigned to the job.
+* `Signal(sig syscall.Signal)` - Send an os signal to the process group assigned to the job.
+
+## Process API
+`Process` is a wrapper around `os/exec.Cmd`
+* `NewProcess(name string, args []string, opts ...Options)` - Returns a new instance of a process given a command name and it's arguments. An optional list of options can also be specified.
+* `Start()` - Starts the process and returns a channel the caller can listen on to get the final status.
+* `Stop()` - Stops the process by sending a `SIGTERM` signal to it's process group.
+* `Signal(sig syscall.Signal)` - Send an os signal to the process group.
+* `Status()` - Get the current status of a process.
+* `IsInitialState()` - true if the process is in the initial state (the process isn't started yet).
+* `IsRunningState()` - true if the process is starting or running.
+* `IsFinalState()` - true if the process is in a final state
+
+## GRPC API
+Please refer to the service proto file for the full definition of the grpc api
+* `rpc Register (JobRequest) returns (JobStatus) {}` - Registers a job without starting it. Returns a unique job id for subsequent use with the other api methods. A status `Initial` is returned to indicate the job has been added.
+* `rpc Start (StartRequest) returns (stream JobStatus) {}` - Start a job that was previously registered using the provided job id. Returns a stream of `JobStatus` for the client to listen to status updates. Returns a `NOT_FOUND` error if the provided job id doesn't exist in the system. It is not an error to call `Start(...)` for a running job or a job in a final state. The stream is closed when the job reaches a final state.
```

Comment:
No need to stream the job status. Users can monitor the job execution by polling the `Status` RPC.
The only streaming component should be the logs

Reply:
Agreed. This will be updated. Originally I thought this could be used as a convenience method but I realize it's unnecessary since there are only about 3 status updates per process and it will especially wasteful for long running processes. Thanks for pointing this out.